words/source/MovePicker.js

```diff
@@ -0,0 +1,283 @@
+  );
+
+  // A specific tile on the board with an index, position and character.
+  class Tile {
+    constructor(charIndex, board) {
+      this._charIndex = charIndex;
+      this._letter = board.letters[charIndex];
+      this._x = charIndex % BOARD_WIDTH;
+      this._y = Math.floor(charIndex / BOARD_HEIGHT);
+    }
+    get charIndex() {
+      return this._charIndex;
+    }
+    get letter() {
+      return this._letter;
+    }
+    get x() {
+      return this._x;
+    }
+    get y() {
+      return this._y;
+    }
+    get toString() {
+      return `[charIndex=${this.charIndex}, letter=${this.letter}, x=${
+        this.x
+      }, y=${this.y}]`;
+    }
+  }
+
+  // A simple dictionary class to encapsulate the set of words that are
+  // considered valid for game purposes.
+  class Dictionary {
+    // Expects words to be provided as a return-delimited string.
+    constructor(words) {
+      // TODO(wkorman): Use a trie for better memory efficiency and ship a
+      // more compact dictionary representation over the wire. Also, strip
+      // proper nouns and words with apostrophes.
+      this._dict = new Set();
+      for (const wordEntry of words.split('\n')) {
+        const trimmedWord = wordEntry.trim();
+        if (trimmedWord.length > 0) this._dict.add(trimmedWord);
+      }
+    }
+    contains(word) {
+      return this._dict.has(word);
+    }
+    get size() {
+      return this._dict.size;
+    }
+  }
+
+  return class extends DomParticle {
+    get template() {
+      return template;
+    }
+    _boardToModels(board, coordinates) {
+      let models = [];
+      // TODO(wkorman): Rework this to use tiles.
+      for (let i = 0; i < board.length; i++) {
+        let x = i % BOARD_WIDTH;
+        let y = Math.floor(i / BOARD_WIDTH);
+        let letterClasses =
+          coordinates.indexOf(`(${x},${y})`) != -1 ? 'selected' : '';
+        models.push({ letter: board[i], index: i, classes: letterClasses });
+      }
+      return models;
+    }
+    _moveToTiles(board, move) {
+      let tiles = [];
+      if (!board || !move || !move.coordinates) return tiles;
+      let tuples = move.coordinates.match(/(\d+,\d+)/g);
+      for (let i = 0; i < tuples.length; i++) {
+        let parts = tuples[i].split(',');
+        let x = parseInt(parts[0]);
+        let y = parseInt(parts[1]);
+        tiles.push(new Tile(y * BOARD_WIDTH + x, board));
+      }
+      return tiles;
+    }
+    _willReceiveProps(props, state) {
+      if (!state.dictionaryLoadingStarted) {
+        let particleRef = this;
+        // TODO(wkorman): Replace this with our own hosted dictionary.
+        const dictionaryName =
+          'https://raw.githubusercontent.com/shaper/shaper.github.io/master/resources/american-english.txt';
+        info(`Loading dictionary [name=${dictionaryName}].`);
+        fetch(dictionaryName).then(response =>
+          response.text().then(text => {
+            let dictionary = new Dictionary(text);
```

Comment:
Not stored in state yet because it isn't being used?

Reply:
Yes, it's being stored and used now, though, excitingly.